src/observer.js

```diff
@@ -0,0 +1,81 @@
+import { parseRootMargin, shallowCompare } from './utils';
+
+export const observerElementsMap = new Map();
+
+export function getPooled(options = {}) {
+    const root = options.root || null;
+    const rootMargin = parseRootMargin(options.rootMargin);
+    const threshold = Array.isArray(options.threshold)
+        ? options.threshold
+        : [options.threshold != null ? options.threshold : 0];
+    const observers = observerElementsMap.keys();
+    let observer;
+    while ((observer = observers.next().value)) {
+        const unmatched =
+            root !== observer.root ||
+            rootMargin !== observer.rootMargin ||
+            shallowCompare(threshold, observer.thresholds);
+
+        if (!unmatched) {
+            return observer;
+        }
+    }
+    return null;
+}
+
+export function findObserverElement(observer, entry) {
+    const elements = observerElementsMap.get(observer);
+    if (elements) {
+        const values = elements.values();
+        let element;
+        while ((element = values.next().value)) {
+            if (element.target === entry.target) {
+                return element;
+            }
+        }
+    }
+    return null;
+}
+
+/**
+ * The Intersection Observer API callback that is called whenever one element,
+ * called the target, intersects either the device viewport or a specified element.
+ * Also will get caled whenever the visibility of the target element changes and
+ * crosses desired amounts of intersection with the root.
+ * @param {array} changes
+ * @param {IntersectionObserver} observer
+ */
+export function callback(changes, observer) {
```

Comment:
`forEach` was in fact smaller, what's the reason to change it for a `for` loop?

Reply:
Speed :) There have been improvements to `forEach` in JS engines, but it still seems to be a lot slower in tight loops. Of course overall this is a micro-level optimization. Only CPU would notice if it were aware of itself.